From 9c6c91b3dc9be2cce15a4231197cbdf603ebe64c Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Marcin=20Wcis=C5=82o?= <marcin.wcislo@conclusive.pl>
Date: Fri, 19 Sep 2025 03:18:33 +0200
Subject: [PATCH] Bluetooth: btsdio: fix use after free bug in btsdio_remove
 due to race condition
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

jira VULN-3967
jira VULN-68015
cve CVE-2023-1989
cve CVE-2023-53145
commit-author Zheng Wang <zyytlz.wz@163.com>
commit 73f7b171b7c09139eb3c6a5677c200dc1be5f318

In btsdio_probe, the data->work is bound with btsdio_work. It will be
started in btsdio_send_frame.

If the btsdio_remove runs with a unfinished work, there may be a race
condition that hdev is freed but used in btsdio_work. Fix it by
canceling the work before do cleanup in btsdio_remove.

	Signed-off-by: Zheng Wang <zyytlz.wz@163.com>
	Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>
(cherry picked from commit 73f7b171b7c09139eb3c6a5677c200dc1be5f318)
	Signed-off-by: Marcin Wcisło <marcin.wcislo@conclusive.pl>
---
 drivers/bluetooth/btsdio.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/drivers/bluetooth/btsdio.c b/drivers/bluetooth/btsdio.c
index d3b0c36609a0..a3ca69685970 100644
--- a/drivers/bluetooth/btsdio.c
+++ b/drivers/bluetooth/btsdio.c
@@ -370,6 +370,7 @@ static void btsdio_remove(struct sdio_func *func)
 	if (!data)
 		return;
 
+	cancel_work_sync(&data->work);
 	hdev = data->hdev;
 
 	sdio_set_drvdata(func, NULL);
-- 
2.47.3

